auditing/order_fulfillment.py

```diff
@@ -0,0 +1,162 @@
+import logging
+import os
+import sys
+
+import MySQLdb
+import pandas
+
+logger = logging.getLogger(__name__)
+
+
+def setup_logging():
+    """ Setup logging.
+
+    Add support for console logging, and avoid truncation by pandas.
+    """
+    msg_format = '%(asctime)s - %(levelname)s - %(message)s'
+    ch = logging.StreamHandler()
+    ch.setFormatter(logging.Formatter(msg_format))
+    ch.setLevel(logging.DEBUG)
+    logger.addHandler(ch)
+    logger.setLevel(logging.DEBUG)
+
+    pandas.set_option('display.max_rows', 500)
+    pandas.set_option('display.max_columns', 500)
+    pandas.set_option('display.width', 1000)
+
+
+ecommerce_db = MySQLdb.connect(host=os.environ['ECOMMERCE_DB_HOST'],
+                               user=os.environ['ECOMMERCE_DB_USER'],
+                               passwd=os.environ['ECOMMERCE_DB_PASSWORD'],
+                               db=os.environ['ECOMMERCE_DB_NAME'])
+
+edxapp_db = MySQLdb.connect(host=os.environ['EDXAPP_DB_HOST'],
+                            user=os.environ['EDXAPP_DB_USER'],
+                            passwd=os.environ['EDXAPP_DB_PASSWORD'],
+                            db=os.environ['EDXAPP_DB_NAME'])
+
+# This is the number of minutes to look back when retrieving orders (e.g. all orders in the last 15 minutes).
+ORDER_WINDOW_START_TIME = int(os.environ.get('ORDER_WINDOW_START_TIME', 20))
+
+
+def get_orders(date_placed):
+    """ Retrieve all completed orders placed after the given date. """
+    sql = """SELECT
+    u.username
+    , o.number
+    , o.date_placed
+    , o.total_excl_tax
+    , modes.value_text AS 'mode'
+    , pav.value_text AS 'course_id'
+FROM
+    order_order o
+    JOIN ecommerce_user u ON (o.user_id = u.id)
+    JOIN order_line ol ON (ol.order_id = o.id)
+    JOIN catalogue_productattributevalue pav ON (pav.product_id = ol.product_id)
+    JOIN catalogue_productattribute pa ON (pa.id = pav.attribute_id AND pa.code = 'course_key')
+    JOIN ( SELECT pav2.product_id, pav2.value_text
+    FROM catalogue_productattributevalue pav2
+    JOIN catalogue_productattribute pa2 ON (pa2.id = pav2.attribute_id AND pa2.code = 'certificate_type')
+    ) AS modes ON (modes.product_id = ol.product_id)
+    LEFT JOIN refund_refundline rl ON (rl.order_line_id = ol.id)
+WHERE
+    o.status = 'Complete'
+    AND o.date_placed > %s
+    AND rl.id IS NULL
+ORDER BY
+    u.username ASC
+    , pav.value_text ASC
+    , o.date_placed ASC;
+"""
+    date_placed = date_placed.strftime('%Y-%m-%d %H:%M:%S')
+    orders = pandas.read_sql_query(sql, ecommerce_db, index_col='number', params=(date_placed,),
+                                   parse_dates=('date_placed',))
+    orders.reset_index(level=0, inplace=True)
+    original_order_count = len(orders)
+
+    # Identify upgrades and remove the honor purchase.
+    # TODO We will need to do the same for credit.
+    grouped = orders.groupby(['username', 'course_id'])
+    for key, row_ids in grouped.groups.iteritems():
```

Comment:
Consider just iterating over the keys and then using `grouped.get_group(key)` to grab a complete DataFrame for that group, which you can then sort in place etc.